lib/slipstream/socket_test.ex

```diff
@@ -0,0 +1,659 @@
+
+  While a client that uses `Slipstream.await_connect/2` can band-aid over
+  these problems with a high enough timeout, clients should be satisfied with
+  receiving successful connection events asynchronously. Clients written in
+  the asynchronous callback style will not be affected.
+
+  Some implementation-level details are glossed over by this testing framework,
+  including that of heartbeat timeouts. If a client sits idle after joining
+  for more than 60 seconds, they will not be terminated due to heartbeat
+  timeout.
+
+  ## Setting up a test
+
+  Another assumption clients typically make of servers is that clients assume
+  there is just _one_ server. A client is not written expecting to hear that it
+  has been connected multiple times for one request to `Slipstream.connect/2`.
+  As such, tests using this module as a case template should be run
+  synchronously.
+
+      defmodule MyApp.MyClientTest do
+        use Slipstream.SocketTest
+
+        ..
+
+  By default, this will start a server session for each test that simulates
+  the current test process as the websocket server connected to the client.
+
+      test "the client sends a push to the server on join", c do
+        accept_connect(MyClient)
+      end
+
+  This server does not run a websocket server. Instead the server is
+  a conceptual server: you may imagine that in each test, you are have control
+  of the remote server and can control the behavior of the server imperatively.
+
+  The `assert_*` and `refute_*` family of macros from this module allow you to
+  make assertions about- and match on values from- requests from the client to
+  the server. The remaining functions allow you to emulate actions on behalf
+  of a hypothetical server.
+
+  ## Timeouts
+
+  The `assert_*` and `refute_*` macros from this module default to ExUnit
+  timeouts. See the ExUnit documentation for more details.
+
+  ## Formatting
+
+  Slipstream exports the `assert_*` and `receive_*` macros as valid
+  `locals_without_parens` in its formatter config. To be able to use these
+  macros without the formatter injecting parentheses, import `:slipstream` in
+  your service's formatter config for `:import_deps`:
+
+      [
+        inputs: ["mix.exs", "{config,lib,test}/**/*.{ex,exs}"],
+        import_deps: [:slipstream],
+        .. # more configuration
+      ]
+  """
+  @moduledoc since: "0.2.0"
+
+  # implementation note:
+  # Getting events to the client process is not difficult.
+  # The author simply must provide the pid or GenServer name. Seizing the
+  # commands being emitted by the client and sent to the connection process is
+  # the name of the game. We do this with two edits to the usual behavior of a
+  # client:
+  #
+  #   1. do not spawn a connection process with `Slipstream.connect/2`
+  #   2. send a `Slipstream.Events.ChannelConnected` with a pid of the current
+  #      test process
+
+  use ExUnit.CaseTemplate
+
+  import Slipstream.Signatures
```

Comment:
I recommend reading through https://hexdocs.pm/slipstream/implementation.html#signatures to learn more about signatures. It's why most things in this module that are `Kernel.send/2`d are wrapped in `event/1` or `command/1`